overlord/hookstate/hookmgr.go

```diff
@@ -295,6 +318,19 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
 		return nil, err
 	}
 
+	// add timeout handling
+	var killTimer *time.Timer
+	if maxRuntime > 0 {
+		killTimer = time.AfterFunc(maxRuntime, func() {
```

Comment:
I like the timer channel and if you follow my systemd advice here we'd just ask systemd to kill the service.